episodes/continuous-integration.md

```diff
@@ -0,0 +1,146 @@
+exercises: 15
+---
+
+:::::: questions
+
+-   What is CI/CD?
+-   Why do we use CI/CD?
+-   What does CI/CD have to do with version control?
+-   What is a CI/CD pipeline?
+-   What is docker and how does it relate to CI/CD?
+    ::::::
+
+:::::: objectives
+
+-   Understand what CI/CD is and why you would use it for your projects.
+-   The relationship between CI/CD and version control
+-   How to build a simple CI/CD pipeline
+    ::::::
+
+GOAL: create awareness or provide a tutorial?
+
+# Continuous Integration and Continuous Deployment / Delivery
+
+## Introduction
+
+Continuous integration is the practice of integrating all your code changes
+into the main branch of a shared source code repository early and often,
+automatically testing each change when you commit or merge them, and
+automatically kicking off a build. -- gitlab.com
+
+Not to be confused with DevOps: CICD is _not_ DevOps, but DevOps requires CICD.
+DevOps is the concept where a team is responsible for the entire life cycle of
+a software product or software component. From development to deployment to
+operating and maintaining it in production. Continuous Integration and
+Continuous Delivery and/or Deployment plays a large role in this. But DevOps is
+not required for leveraging CI/CD.
+
+<https://github.com/resources/articles/devops/ci-cd>
+
+We should distinguis Continuous Integration, Continuous Delivery and Continuous Deployment.
+
+### Continuous Integration
+
+Continuous integration is the practice of integrating all your code changes
+into the main branch of a shared source code repository early and often,
+automatically testing each change when you commit or merge them, and
+automatically kicking off a build. With continuous integration, errors and
+security issues can be identified and fixed more easily, and much earlier in
+the development process. -- gitlab.com
+
+### Continous Delivery
+
+Continuous delivery is a software development practice that works in
+conjunction with CI to automate the infrastructure provisioning and application
+release process. -- gitlab.com
+
+This can be understood as creating a Docker container, creating a PyPi package
+for Python, a jar file for Java or equivalents for programming languages like R
+or C/C++ and Fortran. This will be done in an automated way every time a change
+is pushed to the git repository on github, or gitlab or some other platform.
+
+### Continous Deployment
+
+Continuous deployment enables organizations to deploy their applications
+automatically, eliminating the need for human intervention. -- gitlab.com
+
+When talking about deployment, we mean that the software is running on a server
+and the services it provides are available for consumption by other sofftware
+components. For research software that is the focus of this course, that is
+rarely the case, so we ignore Continous Deployment and focus on Integration and Delivery.
+
+## Why would we want Continuous Integration?
+
+_Todo: needs more text_
```

Comment:
Maybe add here some pointers of how to decide if you need CI or not (for example, if I have a small analysis script, is it worth having CI? and if yes, how would that look like?)